### PR TITLE
refactor(tests): add type hints and clean up

### PR DESCRIPTION
**Pull request by sisp**
I've added type hints to the tests, so they are type-checked by `mypy` as well, and cleaned up the tests a bit, e.g. more consistent setup of temporary template and subproject directories, formatting of template tree creation code for improved readability (IMO), consistent usage of the `pathlib` API instead of the `os`/`os.path` API, etc.

I know it's a quite big PR and probably not much fun to review :see_no_evil: - sorry about that :confused: -, but I felt the tests could use some love :smile_cat:.

### CODE REVIEW

**Comment by yajo** [comment]
Thanks for your efforts! TBH I don't usually like touching tests unless there's an ❌ because there's always risk of breaking stuff. Sometimes they're testing non-obvious corner cases. Unlike with the rest of code, the more tests the better, and being repetitive is no big deal.

That said, the PR looks good. I can tell you know the depths of copier! 😀

I still didn't finish reviewing these 1500 boring lines but i know 1 thing: I wouldn't like to do this again. But I don't see any new mypy checks. Did you add them?

Thanks!

**Comment by sisp** [comment]
I understand your hesitance to touch tests. It's really just a simple refactoring without semantic changes.

I've not added any typing tests, but the added type annotations to the test functions let mypy type-check them, too. mypy doesn't check the typing in bodies of untyped function defs (see [this example](https://mypy-play.net/?mypy=1.0.0&python=3.11&gist=8b2c2f7a307604bc47c581eecc46da92)). Adding `--check-untyped-defs` might help to some extent, too, but I think it's clearer and safer when also the tests are properly type-annotated.

Sorry for the utterly boring review ... 🫣

File: tests/test_migrations.py [hunk 1]
```diff
@@ -72,109 +71,122 @@ def test_migrations_and_tasks(tmp_path: Path):
...
-        copy(src_path=str(template), defaults=True, overwrite=True)
-        answers = json.loads(Path(".copier-answers.yml").read_text())
-        assert answers["_commit"] == "v1"
-        assert answers["best_song"] == "la vie en rose"
-        assert json.loads(Path("songs.json").read_text()) == ["la vie en rose"]
+    copy(src_path=str(src), dst_path=dst, defaults=True, overwrite=True)
+    answers = json.loads((dst / ".copier-answers.yml").read_text())
+    assert answers["_commit"] == "v1"
+    assert answers["best_song"] == "la vie en rose"
+    assert json.loads((dst / "songs.json").read_text()) == ["la vie en rose"]
```
Comment (yajo) [inline tests/test_migrations.py:109]: With this change, you're altering the test.

The previous implementation made sure that `dst_path` defaults to `.`, where the new one doesn't.

Small but important details. Please undo these kind of changes.

File: tests/test_empty_suffix.py [hunk 2]
```diff
@@ -1,57 +1,60 @@
...
+    assert (dst / "pingu.txt").exists()
+    assert (dst / "pingu" / "render_me.txt").exists()
 
     expected = "Hello pingu!"
-    assert (dest / "render_me").read_text() == expected
-    assert (dest / "pingu.txt").read_text() == expected
-    assert (dest / "pingu" / "render_me.txt").read_text() == expected
+    assert (dst / "render_me").read_text() == expected
+    assert (dst / "pingu.txt").read_text() == expected
+    assert (dst / "pingu" / "render_me.txt").read_text() == expected
```
[the remark this answers is not in the document]
Comment (sisp) [reply tests/test_empty_suffix.py:36]: I repsect your position on this diff. But I really believe there is some value in harmonizing the tests a bit, also regarding the recurring creation code for the temporary template and subproject folders as well as the variable names for their paths. It removes variance in the tests where in fact none exists which makes it easier for new contributors to recognize recurring setup code. When several variations are used across the test suite, recognizing semantically identical code is more difficult.

If you really insist on keeping it as it is, I'll do it. But I genuinely think there is value in this harmonization.

File: tests/test_interrupts.py [hunk 3]
```diff
@@ -10,75 +9,66 @@
...
+    ):
         with pytest.raises(CopierAnswersInterrupt) as err:
             worker.run_copy()
-        assert err.value.answers.user == answers
-        assert err.value.template == worker.template
+            assert err.value.answers.user == {
+                "question1": "foobar",
+                "question2": "yosemite",
+            }
+            assert err.value.template == worker.template
```
[the remark this answers is not in the document]
Comment (sisp) [reply tests/test_interrupts.py:74]: OMG, good catch! 🧐:bow: I understand your hesitance to touch tests unnecessarily. This bug is really subtle. I'll revert it of course.

File: tests/test_interrupts.py [hunk 4]
```diff
@@ -10,75 +9,66 @@
...
-    ),
+    "side_effect",
+    [
+        # We override the prompt method from questionary to raise this
+        # exception and expect our surrounding machinery to re-raise
+        # it as a CopierAnswersInterrupt.
+        CopierAnswersInterrupt(Mock(), Mock(), Mock()),
+        KeyboardInterrupt,
+    ],
 )
```
[the remark this answers is not in the document]
Comment (sisp) [reply tests/test_interrupts.py:20]: Thanks! :pray::tada:

File: tests/test_migrations.py (hunk 1 above)
Comment (sisp) [reply tests/test_migrations.py:109]: Ouch, I see and hadn't thought about that! Good catch! :monocle_face::bow:

But one remark: I understand that my change did change the exact test scenario. But I think testing that `dst_path` defaults to `.` should be covered in a dedicated test (which I don't see at the moment) and not only implicitly in some unrelated test because then a change like mine, which was perfectly fine w.r.t. to the scenario under test, coincidentally removed another implicit test. WDYT, @yajo?

**Comment by sisp** [comment]
> Would you get angry at me if I ask you to remove most diffs? :sweat_smile:

No worries, you're the authority of this project and I respect your opinion. No hard feelings. :crossed_fingers: I've tried extrapolating from your inline comments how you'd like this PR to change which has reduced the diff by quite a bit. I'll add some inline comments in a moment where I think there is sufficient value but I'm uncertain whether you'll the changes or where I have a question.

> I see lots of changes that are just stylistic or syntactical. I see some replacing `os.path` for `pathlib.Path`. Others replacing raw strings by `textwrap.dedent`. Others adding some parenthesis to `build_file_tree`... Most of these changes add little value to the code base (mostly when they're found on tests) and introduce big risks when changing just because of yes. OTOH they might fail to test some corner cases.

I completely agree on the cases you've shown that actually had negative impact on the scenarios under test. Regarding stylistic changes (I've also replied to one of your inline comments about this), I think there's some value in stylistic or syntactical harmonization even in the test suite because it reduces or removes variance where in fact none exists, so especially new contributors will find it easier to write better tests because the commonalities like the recurring temporary source/destination setup looks the same everywhere.

File: tests/test_normal_jinja2.py [hunk 5]
```diff
@@ -24,45 +27,52 @@ def test_normal_jinja2(tmp_path_factory):
...
-    todo = (dst / "TODO.txt").read_text()
-    expected = "[[ Guybrush TODO LIST]]\n[# GROG #]\n    - Become a pirate\n"
-    assert todo == expected
+    assert (dst / "TODO.txt").read_text() == dedent(
+        """\
+        [[ Guybrush TODO LIST]]
+        [# GROG #]
+            - Become a pirate
+        """
+    )
```
[the remark this answers is not in the document]
Comment (sisp) [reply tests/test_normal_jinja2.py]: I see. You're right, it was more explicit before. :+1: I'll revert it.

File: tests/test_config.py [hunk 6]
```diff
@@ -427,45 +458,48 @@ def test_user_defaults(tmp_path_factory, user_defaults, data, expected):
...
+    expected_initial: str,
+    expected_updated: str,
+) -> None:
     src, dst = map(tmp_path_factory.mktemp, ("src", "dst"))
     with local.cwd(src):
         build_file_tree(
             {
-                src
-                / "copier.yaml": """\
+                (src / "copier.yaml"): (
```
Comment (sisp) [inline tests/test_config.py:485]: Is it intended that the file path contains `src / ...` here while the current working directory is `src` already due to  `with local.cwd(src):`?

File: tests/test_complex_questions.py [hunk 7]
```diff
@@ -181,55 +191,31 @@ def check_invalid(tui, name, format, invalid_value, help=None, err="Invalid inpu
...
-            [
-                "first",
-                "second",
-                "third",
-            ],
-        )
-    )
+    tui.expect_exact("first")
+    tui.expect_exact("second")
+    tui.expect_exact("third")
```
Comment (sisp) [inline tests/test_complex_questions.py:196]: Do you agree with this change (and the same below)? I think it's better to reduce automation like loops etc. in tests and especially in assertions.

File: tests/test_exclude.py [hunk 8]
```diff
@@ -26,26 +28,39 @@ def test_exclude_recursive_negate(tmp_path):
...
-    src, dst = tmp_path / "src", tmp_path / "dst"
-    build_file_tree({src / "copier.yml": "_exclude: ['*.txt']", src / "aaaa.txt": ""})
+def test_config_exclude(tmp_path_factory: pytest.TempPathFactory) -> None:
+    src, dst = map(tmp_path_factory.mktemp, ("src", "dst"))
+    build_file_tree(
+        {
+            src / "copier.yml": "_exclude: ['*.txt']",
+            src / "aaaa.txt": "",
+        }
+    )
```
Comment (sisp) [inline tests/test_exclude.py:38]: Do you agree with this change (and the other related ones in this file)? It's only stylistic, but I think it's much more readable to use a multiline dict when there is more than file to create.

File: tests/test_prompt.py [hunk 9]
```diff
@@ -154,49 +160,54 @@ def test_copy_default_advertised(tmp_path_factory, spawn, name):
...
+                {
+                    "_envops": BRACKET_ENVOPS,
+                    "_templates_suffix": SUFFIX_TMPL,
+                    "question_1": question_1,
+                    "question_2": {
+                        "type": "yaml",
+                        "default": "something",
+                        "when": question_2_when,
+                    },
+                }
```
Comment (sisp) [inline tests/test_prompt.py:184]: Do you agree with this change? I think that inlining the content of `copier.yml` is more concise and consistent with the rest of the test suite.

File: tests/test_subdirectory.py [hunk 10]
```diff
@@ -155,165 +173,156 @@ def test_update_subdirectory_from_root_path(tmp_path_factory):
...
-            "<<<<<<< before updating\ndownstream version 1\n"
-            "=======\nupstream version 2\n>>>>>>> after updating\n",
+            dedent(
+                """\
+                <<<<<<< before updating
+                downstream version 1
+                =======
+                upstream version 2
+                >>>>>>> after updating
+                """
```
Comment (sisp) [inline tests/test_subdirectory.py:190]: Do you agree with using a multiline string for the inline merge conflict? I think it's much more readable. The line breaks are implicit, yes, but it's much more obvious what the merge conflict syntax is.

File: tests/test_templated_prompt.py [hunk 11]
```diff
@@ -236,14 +244,19 @@ def test_templated_prompt_builtins(tmp_path_factory):
...
 @pytest.mark.parametrize(
     "questions, raises, returns",
-    (
+    [
         ({"question": {"default": "{{ not_valid }}"}}, None, ""),
         ({"question": {"help": "{{ not_valid }}"}}, None, "None"),
         ({"question": {"type": "{{ not_valid }}"}}, InvalidTypeError, "None"),
         ({"question": {"choices": ["{{ not_valid }}"]}}, None, "None"),
-    ),
+    ],
```
Comment (sisp) [inline tests/test_templated_prompt.py:252]: There are a few of these minor changes in this PR. I think using a list of parameters is semantically better than using a tuple because in my understanding lists have the semantics of being collections while tuples have the semantics of ordered multi-value data. It's a minor distinction and both work fine technically. Also, [pytest's documentation](https://docs.pytest.org/en/stable/example/parametrize.html) uses a list if that's worth anything. :wink:

File: tests/test_config.py [hunk 12]
```diff
@@ -36,7 +36,7 @@ def git_init(message="hello world") -> None:
 
 
 def test_config_data_is_loaded_from_file() -> None:
-    tpl = Template(str(Path("tests", "demo_data")))
+    tpl = Template("tests/demo_data")
```
Comment (sisp) [inline tests/test_config.py:39]: I've reverted these kinds of changes because I think Copier handles POSIX paths fine both on Unix and Windows due to `pathlib.Path()` usage.

I tested this with `pathlib`:

```python
>>> pathlib.PureWindowsPath("a/b").parts
('a', 'b')
```

And `pathlib.Path()` seems to delegate the OS-specific implementation to the respective classes.

File: tests/test_config.py (hunk 6 above)
Comment (yajo) [reply tests/test_config.py:485]: probably just an overlook

File: tests/test_complex_questions.py (hunk 7 above)
Comment (yajo) [reply tests/test_complex_questions.py:196]: These are just 2 ways to express the same thing, so I'd keep it as it was.

File: tests/test_config.py (hunk 6 above)
Comment (sisp) [reply tests/test_config.py:485]: So I'll remove the `src / ` part, it shouldn't change the behavior of that test.

File: tests/test_empty_suffix.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (yajo) [reply tests/test_empty_suffix.py:36]: It's not an easy decision to take. On one hand, I obviously like prettier code. But OTOH, I already found 2 issues with the refactor, and the old motto "if something ain't broken, don't fix it" comes to my mind. I'm afraid that with the refactor we're introducing more non-obvious bugs, or at least skipping some other non-obvious corner cases that are being implicitly tested.

All my life I considered tests the only part of code that is "write and forget". I don't actually "maintain" tests. I just fix them when they're broken. And those breaks help me to know that the next release will be a major one, because something is broken. Also, writing tests is boring enough, so when a contributor contributes a test that is not diverging too much from the somehow modern way of testing copier, I just accept that. I mean: as long as the test is testing what it must test and goes green, it's mostly OK for me.

Putting all things in balance, sadly I think stability is weightier than beauty, so let me gladly ask you to please start reverting most of the diff. Let me do more comments on your specific questions below.

File: tests/test_prompt.py (hunk 9 above)
Comment (yajo) [reply tests/test_prompt.py:184]: Putting risk/value into balance, I think it's better to undo it.

File: tests/test_exclude.py (hunk 8 above)
Comment (yajo) [reply tests/test_exclude.py:38]: Let's undo it too.

File: tests/test_migrations.py (hunk 1 above)
Comment (yajo) [reply tests/test_migrations.py]: Of course, specific tests are better. But as long as we don't touch them, they won't add too much. Just undo the change; that'll be the most practical solution.

File: tests/test_subdirectory.py (hunk 10 above)
Comment (yajo) [reply tests/test_subdirectory.py:190]: OK, let's keep this one.

File: tests/test_templated_prompt.py (hunk 11 above)
Comment (yajo) [reply tests/test_templated_prompt.py:252]: Let's undo this change too please.

File: tests/test_config.py (hunk 12 above)
Comment (yajo) [reply tests/test_config.py:39]: let's undo these too please.

File: tests/test_exclude.py (hunk 8 above)
Comment (sisp) [reply tests/test_exclude.py]: Oh well ... Done. :shrug:

File: tests/test_prompt.py (hunk 9 above)
Comment (sisp) [reply tests/test_prompt.py]: Oh well ... Done. :shrug:

File: tests/test_templated_prompt.py (hunk 11 above)
Comment (sisp) [reply tests/test_templated_prompt.py]: Oh well ... Done. :shrug:

File: tests/test_empty_suffix.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sisp) [reply tests/test_empty_suffix.py]: Oh well ... Done. :shrug:

**Comment by sisp** [comment]
I've reverted some more changes. Would you like to take a look again, @yajo?

> On one hand, I obviously like prettier code. But OTOH, I already found 2 issues with the refactor, and the old motto "if something ain't broken, don't fix it" comes to my mind. I'm afraid that with the refactor we're introducing more non-obvious bugs, or at least skipping some other non-obvious corner cases that are being implicitly tested.
>
> All my life I considered tests the only part of code that is "write and forget". I don't actually "maintain" tests. I just fix them when they're broken. And those breaks help me to know that the next release will be a major one, because something is broken. Also, writing tests is boring enough, so when a contributor contributes a test that is not diverging too much from the somehow modern way of testing copier, I just accept that. I mean: as long as the test is testing what it must test and goes green, it's mostly OK for me.

I'm not sure how much further I want to take this discussion or just give in. :confused: Of course I understand the desire to keep the tests functional and the risk of losing that. But I disagree with tests code being "write and forget" because test code may evolve in similar ways as the main code does. And deliberately not maintaining test code contributes to technical debt. For instance, a project may have been started by a junior developer, which may have led to a bad tests organization, bad test scope, etc.; or the reviews in the early days of a project weren't as thorough as later; or certain testing patterns (setup/teardown code, naming conventions, etc.) became apparent only at a later point of the project. Those are all perfectly valid reasons to _maintain_, i.e. continuously refactor and clean up, a test suite because technical debt should be kept in check. While reading on this topic a bit, advice in several places was to never refactor the test suite _and_ change the main code which makes total sense and I haven't touched both here. Of course, there's no 100% guarantee, but the combination of leaving the main code untouched, using code quality metrics like code coverage as safeguards, using a linter, and thorough code review should minimize that risk. And especially purely stylistic changes, some of which I've even reverted by now, are obviously not problematic.

**Review by yajo — APPROVED** [review]
All those points have pros and cons. There's no silver bullet here.

For example, stylistic changes may seem more beautiful to some people and uglier to some others. They may make the code look better, but they also make `git blame` harder.

There's a difference between test _maintenance_ (which is important) and test _refactor_ (which is risky). For example, a test might have a bottleneck and make CI go slow. Or a library used for tests might introduce some breaking change and needs to update some tests. Or some app change involves a change in a test. That's just maintenance and is a good thing.

What I think is: what adds more value? To refactor a test, or to add a new one? And what is more risky?

In any case, let's merge this one as it seems to be OK now. Thanks for your help!
